ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartCompleteUploadHandler.java

```diff
@@ -382,12 +383,27 @@ private boolean isRetriable(Throwable throwable) {
     List<ChunkInfo> getChunksToStitch(CompleteMultipartUpload completeMultipartUpload) throws RestServiceException {
       // Get parts in order from CompleteMultipartUpload, deserialize each part id to get data chunk ids.
       List<ChunkInfo> chunkInfos = new ArrayList<>();
+      Set<String> eTags = new HashSet<>();
       try {
         // sort the list in order
         List<Part> sortedParts = Arrays.asList(completeMultipartUpload.getPart());
         Collections.sort(sortedParts, Comparator.comparingInt(Part::getPartNumber));
         String reservedMetadataId = null;
         for (Part part : sortedParts) {
+          // Validate part number: must be a non-negative integer and within the range 0-10,000
+          if (part.getPartNumber() < 0 || part.getPartNumber() > 10000 || part.getPartNumber() != Math.floor(part.getPartNumber())) {
```

Comment:
Also verify the size of the list. That should be less than 10000. Don't just verify part numbers. With this fix I can trick the system by passing a million entries all with part numbers less than 10000.

And this patch is adding two fixes. I'd recommend splitting this into two. One for etag. One for part numbers. For part numbers we need to be careful and ensure we are restricting customers from uploading large blobs.